pithos/pithos.py

```diff
@@ -818,38 +866,36 @@ def check_if_song_is_ad(self):
                 logging.warning('dur_stat is False. The assumption that duration is available once the stream-start messages feeds is bad.')
 
     def on_gst_buffering(self, bus, message):
-        # per GST documentation:
-        # Note that applications should keep/set the pipeline in the PAUSED state when a BUFFERING
-        # message is received with a buffer percent value < 100 and set the pipeline back to PLAYING
-        # state when a BUFFERING message with a value of 100 percent is received.
-
-        # 100% doesn't mean the entire song is downloaded, but it does mean that it's safe to play.
-        # trying to play before 100% will cause stuttering.
-        percent = message.parse_buffering()
-        logging.debug("Buffering (%i%%)", percent)
-
-        if percent < 100:
-            # If our previous buffer was at 100, but now it's < 100,
-            # then we should pause until the buffer is full.
-            if self.player_status.buffer_percent == 100:
-                logging.debug("Buffer underrun. Pausing pipeline")
-                self.player.set_state(Gst.State.PAUSED)
-                self.player_status.began_buffering = time.time()
+        # If the pipeline signals that it is buffering set the player to PseudoGst.BUFFERING
+        # (which is an alias to Gst.State.PAUSED). During buffering if the user goes to Pause
+        # or Play(an/or back again) go though all the motions but don't actaully change the
+        # player's state to the desired state until buffering has completed. The player only
+        # cares about the actual state, the rest of Pithos only cares about our buffer_recovery
+        # state, the state we *wish* we were in.
+
+        if self.query_buffer():
+            if self._current_state is not PseudoGst.BUFFERING:
+                logging.debug("Buffer underrun")
+                if self._set_player_state(PseudoGst.BUFFERING):
+                    logging.debug("Pausing pipeline")
+                    self.update_song_row()
         else:
-            if self.playing is None: # Not playing but waiting to
-                logging.debug("Buffer 100%. Song starting")
-                self.play()
-            elif self.playing:
-                logging.debug("Buffer recovery. Restarting pipeline")
-                self.player.set_state(Gst.State.PLAYING)
-            else:
-                logging.debug("Buffer recovery. User paused")
-            # Tell everyone to update their clocks after we're done buffering or
-            # in case it takes a while after the song-changed signal for actual playback to begin.
-            self.emit('buffering-finished', self.query_position() or 0)
-            self.player_status.began_buffering = None
-        self.player_status.buffer_percent = percent
-        self.update_song_row()
+            if self._current_state is PseudoGst.BUFFERING:
```

Comment:
I guess I don't understand your question here. self._buffer_recovery_state can only be PseudoGst.STOPPED, PseudoGst.PLAYING or PseudoGst.PAUSED. That's why there's a couple elif's.